exchanges/exchange.go

```diff
@@ -1752,3 +1751,64 @@ func (b *Base) MatchSymbolCheckEnabled(symbol string, a asset.Item, hasDelimiter
 func (b *Base) IsPairEnabled(pair currency.Pair, a asset.Item) (bool, error) {
 	return b.CurrencyPairs.IsPairEnabled(pair, a)
 }
+
+// Bootstrap function allows for exchange authors to supplement or override common startup actions
+// If the result of exchange.Bootstrap is false it will not perfom any other actions.
+// If it returns true, or is not implemented by the exchange, it will:
+// * Print debug startup information
+// * UpdateOrderExecutionLimits
+// * UpdateTradablePairs
+func Bootstrap(ctx context.Context, b IBotExchange) {
+	if continueBootstrap := b.Bootstrap(ctx); !continueBootstrap {
+		return
+	}
+
+	if b.IsVerbose() {
+		wsURL := ""
+		wsEnabled := false
+		if w, err := b.GetWebsocket(); err != nil {
+			wsURL = w.GetWebsocketURL()
+			wsEnabled = w.IsEnabled()
+		}
+		log.Debugf(log.ExchangeSys, "%s Websocket: %s. (url: %s).\n", b.GetName(), common.IsEnabled(wsEnabled), wsURL)
+		b.PrintEnabledPairs()
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		a := b.GetAssetTypes(true)
+		for x := range a {
+			if err := b.UpdateOrderExecutionLimits(ctx, a[x]); err != nil && !errors.Is(err, common.ErrNotYetImplemented) {
+				log.Errorf(log.ExchangeSys, "%s failed to set exchange order execution limits. Err: %v", b.GetName(), err)
+			}
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if b.GetEnabledFeatures().AutoPairUpdates {
+			if err := b.UpdateTradablePairs(ctx, false); err != nil {
+				log.Errorf(log.ExchangeSys, "%s failed to update tradable pairs. Err: %s", b.GetName(), err)
+			}
+		}
+	}()
+
+	wg.Wait()
+}
+
+// Bootstrap is a fallback method for exchange startup actions
+// Exchange authors should overide this if they wysh to customise startup actions
+// Return true to all default Bootstrap actions to occur afterwars
+// or false to signal that no further bootstrapping should occur
+func (b *Base) Bootstrap(_ context.Context) (continueBootstrap bool) {
```

Comment:
Can we return an error for this method, so we don't need to log everything wrapper side if implemented?

Reply:
I thought the same thing myself more than once.
If we both thought it, Let's make it so.
On it like a car bonnet.

Reply:
Fixed abac8c89

Note Well: Behaviour change; We will now return errors all the way up if Bootstrap errors.
I think this is a good thing.